server/src/main/java/org/opensearch/index/translog/transfer/TranslogCheckpointTransferSnapshot.java

```diff
@@ -0,0 +1,148 @@
+            .append(" ]")
+            .toString();
+    }
+
+    /**
+     * Builder for {@link TranslogCheckpointTransferSnapshot}
+     */
+    public static class Builder {
+        private final long primaryTerm;
+        private final long generation;
+        private final List<TranslogReader> readers;
+        private final Function<Long, String> checkpointGenFileNameMapper;
+        private final Path location;
+
+        public Builder(
+            long primaryTerm,
+            long generation,
+            Path location,
+            List<TranslogReader> readers,
+            Function<Long, String> checkpointGenFileNameMapper
+        ) {
+            this.primaryTerm = primaryTerm;
+            this.generation = generation;
+            this.readers = readers;
+            this.checkpointGenFileNameMapper = checkpointGenFileNameMapper;
+            this.location = location;
+        }
+
+        public TranslogCheckpointTransferSnapshot build() throws IOException {
+            final List<Long> generations = new LinkedList<>();
+            long highestGeneration = Long.MIN_VALUE;
+            long highestGenPrimaryTerm = Long.MIN_VALUE;
+            long lowestGeneration = Long.MAX_VALUE;
+            long highestGenMinTranslogGeneration = Long.MIN_VALUE;
+            TranslogCheckpointTransferSnapshot translogTransferSnapshot = new TranslogCheckpointTransferSnapshot(
+                primaryTerm,
+                generation,
+                readers.size()
+            );
+            for (TranslogReader reader : readers) {
+                final long readerGeneration = reader.getGeneration();
+                final long readerPrimaryTerm = reader.getPrimaryTerm();
+                final long minTranslogGeneration = reader.getCheckpoint().getMinTranslogGeneration();
+                final long checkpointGeneration = reader.getCheckpoint().getGeneration();
+                Path translogPath = reader.path();
+                Path checkpointPath = location.resolve(checkpointGenFileNameMapper.apply(readerGeneration));
+                generations.add(readerGeneration);
+                translogTransferSnapshot.add(
+                    new TranslogFileSnapshot(readerPrimaryTerm, readerGeneration, translogPath),
+                    new CheckpointFileSnapshot(readerPrimaryTerm, checkpointGeneration, minTranslogGeneration, checkpointPath)
+                );
+                if (readerGeneration > highestGeneration) {
+                    highestGeneration = readerGeneration;
+                    highestGenMinTranslogGeneration = minTranslogGeneration;
+                    highestGenPrimaryTerm = readerPrimaryTerm;
+                }
+                lowestGeneration = Math.min(lowestGeneration, readerGeneration);
+            }
+            translogTransferSnapshot.setMinTranslogGeneration(highestGenMinTranslogGeneration);
+
+            assert this.primaryTerm == highestGenPrimaryTerm : "inconsistent primary term";
```

Comment:
Again with me complaining about asserts :) but I would enforce these constraints unconditionally. Performance doesn't seem like a concern since the work is done unconditionally. Same applies for line 50.

I won't block on this issue if you strongly prefer asserts, but regardless we should have unit tests for these constraints.

Reply:
I'll stick with this for now and make changes as needed once we added tests for this.